### ISSUE TITLE
Interceptor not working in angular 2 feature module ?

### ISSUE THREAD

**Issue by figuerres** [gund/ng-http-interceptor#120]
i have tried a few times and it seems that in a feature module the interectptor does not get installed or called.

in my root level app.module  it works as expected and i can see headers added to calls and logging in my console log when i call a rest api service.
but when i have a  service in a feature module the interectptor is not working.
i tried adding the interecptor to the feature module and that does not add it.

what do i need to do to debug this or make this work for my entire application and all feature modules ?

**Comment by gund** [comment]
As I understand your `FeatureModule` imported somewhere in app in `AppModule`, right?
For this you do not need to do any extra steps, just import `HttpInterceptorModule` in your `AppModule` and that should be enough.

Basically the idea that you need to import this module only once in the root module of your app.

**Comment by figuerres** [comment]
@gund , please review the angular 2 documentation Here: https://angular.io/docs/ts/latest/guide/ngmodule.html#!#feature-modules

in short a "feature Module" is an Ng Module that contains a set of components.
the app module does not import the feature module as such. the router navigates to the module.
so each feature module is almost like a seperate app.

the router gets a section like:
` {
                path: 'track',
                loadChildren: 'app/track/track.module#TrackModule',
 },`

that tells the angular 2 framework to load and run the module.

that module has it's ngModule  that has it's own imports and providers.

so the service call made from a service in that module is not interecpted by the interectptor that is in the app component.

a service call from a service in the app module is handled as expected.

i then tried to add the imports and logic to the feature module but it did not seem to work and no header is added and no logging that the interceptor is called.

right now i am going to try the   HTTP_INTERCEPTOR_PROVIDER,  and see if it works in the feature module.

**Comment by gund** [comment]
It seems you misunderstood a concept of a _feature module_: It's a module that might encapsulate or expose specific features/functionality and *should* be imported into root module in order to share same _injector_.

Have a look at this key concepts from the link you sent:
* The root module and the feature module share the same execution context. They share the same dependency injector which means the services in one module are available to all.
* We boot the root module to launch the app; we import [into root module] a feature module to extend the app.

And what you are trying to do is to have duplicate of same HttpInterceptor itself, that's why it does not work in your feature module.
And if you will just import `HttpInterceptorModule` or it's providers - it will just create second instances of the module (which might still work, but it's not right and might cause some undefined behaviour).

And by the way providers (like HTTP_INTERCEPTOR_PROVIDER) will be removed starting from v2
to utilize only one consistent way of using library.

**Comment by figuerres** [comment]
Looks like i had an error and a race condition .. now fixed.

but please do look at this plunk from the angular documentation on feature modules:
"Angular Modules (Final)" plunker 
has feature modules called "crisis" and "hero"  but the app.module.ts does not import them.  the app.component.ts does not import them but has router links to them.
app.routing.module  has urls for them but does not import them.

but what you said about the shared context is true, i did find that the interecptor is now working on my feature module as you expected.
i had to not import any of the interceptor in my feature module and fix the bug i had in other code.

thanks for your help!

**Comment by gund** [comment]
Ok I see now what you meant =)

Glad you resolved an issue.
